### ISSUE TITLE
MacOS Mojave : Unable to discover device(s), cannot find U2F device

### ISSUE THREAD

**Issue by mfrancois** [Yubico/pam-u2f#124]
Hi,
I have an issue to use U2F with titan google key.  I am on MacOS 10.14.5 (18F132) Mojave.
Her my installation :
```
# Install PAM U2F Module
brew install pam-u2f

# Generate a new authorization mapping file
mkdir -p ~/.config/u2f  
# >>> Insert U2F Token
pamu2fcfg > ~/.config/u2f/keys  
# >>> Press button on U2F token
# >>> Remove token
# >>> Insert second U2F token
pamu2fcfg -n >> ~/.config/u2f/keys  
# >>> Press button on second U2F token
# >>> Remove token

# Edit PAM configuration files
sudo nano /etc/pam.d/screensaver  
### Add a line with:
auth       sufficient  /usr/local/opt/pam-u2f/lib/pam/pam_u2f.so debug debug_file=/Users/mfrancois/.config/u2f/debug.log  authfile=/Users/mfrancois/.config/u2f/keys

sudo nano /etc/pam.d/authorization  
### Add a line with:
auth       sufficient  /usr/local/opt/pam-u2f/lib/pam/pam_u2f.so debug debug_file=/Users/mfrancois/.config/u2f/debug.log  authfile=/Users/mfrancois/.config/u2f/keys

```

I made some test with 

```
u2f-server -aregister -ohttp://demo.yubico.com   -i http://demo.yubico.com -k keyhandle.dat -p userkey.dat
```

I past the json result on 

```
u2f-host -aregister -o http://demo.yubico.com
```

I past the json result on previous one an I have **Registration successful**
I test the key on demo website and that work well to. 

But when I would like use on mac authentification the debug let me know **Unable to discover device(s), cannot find U2F device**

```
debug(pam_u2f): util.c:167 (get_devices_from_authfile): Length of key number 2 is 65
debug(pam_u2f): util.c:194 (get_devices_from_authfile): Found 2 device(s) for user mfrancois
debug(pam_u2f): pam-u2f.c:340 (pam_sm_authenticate): Using file '/var/run/user/0/pam-u2f-authpending' for emitting touch request notifications
debug(pam_u2f): util.c:271 (do_authentication): Unable to discover device(s), cannot find U2F device
debug(pam_u2f): pam-u2f.c:371 (pam_sm_authenticate): do_authentication returned -2
debug(pam_u2f): pam-u2f.c:410 (pam_sm_authenticate): done. [authentication error]
debug(pam_u2f): pam-u2f.c:99 (parse_cfg): called.
debug(pam_u2f): pam-u2f.c:100 (parse_cfg): flags 0 argc 3
debug(pam_u2f): pam-u2f.c:102 (parse_cfg): argv[0]=debug
debug(pam_u2f): pam-u2f.c:102 (parse_cfg): argv[1]=debug_file=/Users/mfrancois/.config/u2f/debug.log
debug(pam_u2f): pam-u2f.c:102 (parse_cfg): argv[2]=authfile=/Users/mfrancois/.config/u2f/keys
debug(pam_u2f): pam-u2f.c:104 (parse_cfg): max_devices=0
debug(pam_u2f): pam-u2f.c:105 (parse_cfg): debug=1
debug(pam_u2f): pam-u2f.c:106 (parse_cfg): interactive=0
debug(pam_u2f): pam-u2f.c:107 (parse_cfg): cue=0
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): nodetect=0
debug(pam_u2f): pam-u2f.c:109 (parse_cfg): manual=0
debug(pam_u2f): pam-u2f.c:110 (parse_cfg): nouserok=0
debug(pam_u2f): pam-u2f.c:111 (parse_cfg): openasuser=0
debug(pam_u2f): pam-u2f.c:112 (parse_cfg): alwaysok=0
debug(pam_u2f): pam-u2f.c:113 (parse_cfg): authfile=/Users/mfrancois/.config/u2f/keys
debug(pam_u2f): pam-u2f.c:114 (parse_cfg): authpending_file=(null)
debug(pam_u2f): pam-u2f.c:115 (parse_cfg): origin=(null)
debug(pam_u2f): pam-u2f.c:116 (parse_cfg): appid=(null)
debug(pam_u2f): pam-u2f.c:117 (parse_cfg): prompt=(null)
debug(pam_u2f): pam-u2f.c:169 (pam_sm_authenticate): Origin not specified, using "pam://mbp-de-maxime-2.lan"
debug(pam_u2f): pam-u2f.c:181 (pam_sm_authenticate): Appid not specified, using the same value of origin (pam://mbp-de-maxime-2.lan)
debug(pam_u2f): pam-u2f.c:192 (pam_sm_authenticate): Maximum devices number not set. Using default (24)
debug(pam_u2f): pam-u2f.c:210 (pam_sm_authenticate): Requesting authentication for user mfrancois
debug(pam_u2f): pam-u2f.c:221 (pam_sm_authenticate): Found user mfrancois
debug(pam_u2f): pam-u2f.c:222 (pam_sm_authenticate): Home directory for mfrancois is /Users/mfrancois
debug(pam_u2f): pam-u2f.c:271 (pam_sm_authenticate): Using authentication file /Users/mfrancois/.config/u2f/keys
debug(pam_u2f): util.c:105 (get_devices_from_authfile): Authorization line: mfrancois:H43INPz6_XLBOHKqD0vyIGNxIeUX--mjWNCgMCJtZpatpW9pCbEtJ7hjtYoe_yRacYRriqd_y0s-YsQDuSOmxw,0442ddb92f941cf73ca3851d535ed70126d051edc51aa899345179671c0a09d70e98bab62b8f4da5be78abe4ea3af766194f0df4e5c0479ac6138c8d8cd80402b2:aBGhbYSZv6eqeIAYs4gjhYAEj2R_dMbdH2yN5eHuVpzFFSEjOivpqhLPNbXKgYobI-ZNIPtJl9X31RlMEgFOAg,04acd50d7231c61ed4ae560e92e9362a6fa8fa0ae3a837508949b2f60fda2a940336b6e3267b1230f3f5077ff1fdba8ba6765ae30ec6dcd0cbd589d03b630892d2
debug(pam_u2f): util.c:110 (get_devices_from_authfile): Matched user: mfrancois
debug(pam_u2f): util.c:137 (get_devices_from_authfile): KeyHandle for device number 1: H43INPz6_XLBOHKqD0vyIGNxIeUX--mjWNCgMCJtZpatpW9pCbEtJ7hjtYoe_yRacYRriqd_y0s-YsQDuSOmxw
debug(pam_u2f): util.c:156 (get_devices_from_authfile): publicKey for device number 1: 0442ddb92f941cf73ca3851d535ed70126d051edc51aa899345179671c0a09d70e98bab62b8f4da5be78abe4ea3af766194f0df4e5c0479ac6138c8d8cd80402b2
debug(pam_u2f): util.c:167 (get_devices_from_authfile): Length of key number 1 is 65
debug(pam_u2f): util.c:137 (get_devices_from_authfile): KeyHandle for device number 2: aBGhbYSZv6eqeIAYs4gjhYAEj2R_dMbdH2yN5eHuVpzFFSEjOivpqhLPNbXKgYobI-ZNIPtJl9X31RlMEgFOAg
debug(pam_u2f): util.c:156 (get_devices_from_authfile): publicKey for device number 2: 04acd50d7231c61ed4ae560e92e9362a6fa8fa0ae3a837508949b2f60fda2a940336b6e3267b1230f3f5077ff1fdba8ba6765ae30ec6dcd0cbd589d03b630892d2
debug(pam_u2f): util.c:167 (get_devices_from_authfile): Length of key number 2 is 65
debug(pam_u2f): util.c:194 (get_devices_from_authfile): Found 2 device(s) for user mfrancois
debug(pam_u2f): pam-u2f.c:340 (pam_sm_authenticate): Using file '/var/run/user/0/pam-u2f-authpending' for emitting touch request notifications
debug(pam_u2f): util.c:271 (do_authentication): Unable to discover device(s), cannot find U2F device
debug(pam_u2f): pam-u2f.c:371 (pam_sm_authenticate): do_authentication returned -2
debug(pam_u2f): pam-u2f.c:410 (pam_sm_authenticate): done. [authentication error]
```

**Comment by a-dma** [comment]
I've tried to reproduce the issue, and as far as I can see it seems to be a bit intermittent, for example I sometimes get failures also with `pamu2fcfg`, except that tool retries opening the device a couple of times and it eventually succeeds, same for `u2f-host -a register`.

I believe this is an issue with the HID layer. Would you be able to build and test the latest version on master? That one uses `libfido2` instead of `libu2f-{server,host}` and re-implements part of the HID logic.

**Comment by daysonhw** [comment]
> I've tried to reproduce the issue, and as far as I can see it seems to be a bit intermittent, for example I sometimes get failures also with `pamu2fcfg`, except that tool retries opening the device a couple of times and it eventually succeeds, same for `u2f-host -a register`.
> 
> I believe this is an issue with the HID layer. Would you be able to build and test the latest version on master? That one uses `libfido2` instead of `libu2f-{server,host}` and re-implements part of the HID logic.


I have tried with 1.0.8 with my two device, they both have the problem. pamu2fcfg works fine. but the debuger still says cannot find U2F device

**Comment by a-dma** [comment]
Would you be able to try `master`?

**Comment by daysonhw** [comment]
> Would you be able to try `master`?

I stuck with missing 'libcrypto', and I don't know how to use pkg-config to fix it.

**Comment by a-dma** [comment]
First off, thanks for testing.

`libcrypto` is just OpenSSL. You can install it from `brew` and at that point `pkg-config` should pick it up for you. If it doesn't you can help it by setting `PKG_CONFIG_PATH` to the directory containing the `.pc` file.

**Comment by daysonhw** [comment]
> First off, thanks for testing.
> 
> `libcrypto` is just OpenSSL. You can install it from `brew` and at that point `pkg-config` should pick it up for you. If it doesn't you can help it by setting `PKG_CONFIG_PATH` to the directory containing the `.pc` file.

Thank you! That works
but after building, I only get the pam-u2f.lo and pam-u2f.la. I don't see the pam-u2f.so in the root or some where. Can you give me some advice?

**Comment by a-dma** [comment]
Have a look inside the `.libs` directory which Autotools conveniently hide from you.

**Comment by daysonhw** [comment]
> Have a look inside the `.libs` directory which Autotools conveniently hide from you.

Succeed!
After I put it in the pam.d. After I enter my password the device flickered, but it did't give me the time to click the button before reject me.

Here is the pam-u2f.so
https://drive.google.com/open?id=1OfI7iVQfN9-TQkbjmfkSK7HxxnOumtoG

**Comment by daysonhw** [comment]
debug(pam_u2f): pam-u2f.c:105 (parse_cfg): called.
debug(pam_u2f): pam-u2f.c:106 (parse_cfg): flags 0 argc 3
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): argv[0]=debug
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): argv[1]=debug_file=/Users/dayson/.config/u2f/debug.log
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): argv[2]=authfile=/Users/dayson/.config/u2f/keys
debug(pam_u2f): pam-u2f.c:110 (parse_cfg): max_devices=0
debug(pam_u2f): pam-u2f.c:111 (parse_cfg): debug=1
debug(pam_u2f): pam-u2f.c:112 (parse_cfg): interactive=0
debug(pam_u2f): pam-u2f.c:113 (parse_cfg): cue=0
debug(pam_u2f): pam-u2f.c:114 (parse_cfg): nodetect=0
debug(pam_u2f): pam-u2f.c:115 (parse_cfg): userpresence=0
debug(pam_u2f): pam-u2f.c:116 (parse_cfg): userverification=0
debug(pam_u2f): pam-u2f.c:117 (parse_cfg): pinverification=0
debug(pam_u2f): pam-u2f.c:118 (parse_cfg): manual=0
debug(pam_u2f): pam-u2f.c:119 (parse_cfg): nouserok=0
debug(pam_u2f): pam-u2f.c:120 (parse_cfg): openasuser=0
debug(pam_u2f): pam-u2f.c:121 (parse_cfg): alwaysok=0
debug(pam_u2f): pam-u2f.c:123 (parse_cfg): authfile=/Users/dayson/.config/u2f/keys
debug(pam_u2f): pam-u2f.c:125 (parse_cfg): authpending_file=(null)
debug(pam_u2f): pam-u2f.c:126 (parse_cfg): origin=(null)
debug(pam_u2f): pam-u2f.c:127 (parse_cfg): appid=(null)
100 fewer lines
debug(pam_u2f): pam-u2f.c:366 (pam_sm_authenticate): Unable to emit 'authentication started' notification by opening the file '/var/run/user/0/pam-u2f-authpending', (No such file or directory)
debug(pam_u2f): util.c:527 (do_authentication): Device max index is 1
debug(pam_u2f): util.c:560 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:581 (do_authentication): Key handle: syF2PvdaZEupAook8N7X2lkEpxtPKDcceHxpNqvmKYZj2jajmHqKg9CcqCK1h4XcnUkJIQO2MugSj82E231CmouQPJzKRsvsNt/td1rnLgY==
debug(pam_u2f): util.c:673 (do_authentication): Challenge: ZdrTc7DYB8EhTV/ZnYMoXI3n6nIhxjoESnH9IY/oCOo=
debug(pam_u2f): util.c:422 (get_authenticators): Working with 1 authenticator(s)
debug(pam_u2f): util.c:426 (get_authenticators): Checking whether key exists in authenticator 0
debug(pam_u2f): util.c:436 (get_authenticators): Authenticator path: IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/GP17@7,1/IOPP/XHC0@0,3/XHC0@60000000/PRT3@60300000/USB2.0 Hub@60300000/AppleUSB20Hub@60300000/AppleUSB20HubPort@60310000/ePass FIDO@60310000/IOUSBHostInterface@0/IOUSBHostHIDDevice@60310000,0
debug(pam_u2f): util.c:463 (get_authenticators): Found key in authenticator 0
debug(pam_u2f): pam-u2f.c:391 (pam_sm_authenticate): do_authentication returned -2
debug(pam_u2f): pam-u2f.c:430 (pam_sm_authenticate): done. [authentication error]
debug(pam_u2f): pam-u2f.c:105 (parse_cfg): called.


debug(pam_u2f): pam-u2f.c:366 (pam_sm_authenticate): Unable to emit 'authentication started' notification by opening the file '/var/run/user/0/pam-u2f-authpending', (No such file or directory)
debug(pam_u2f): util.c:527 (do_authentication): Device max index is 1
debug(pam_u2f): util.c:560 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:581 (do_authentication): Key handle: syF2PvdaZEupAook8N7X2lkEpxtPKDcceHxpNqvmKYZj2jajmHqKg9CcqCK1h4XcnUkJIQO2MugSj82E231CmouQPJzKRsvsNt/td1rnLgY==
debug(pam_u2f): util.c:673 (do_authentication): Challenge: IR3YPMYjtlTDAM2QGHe2caq/zIsJCenAsZOJKz1Zi2s=
debug(pam_u2f): util.c:422 (get_authenticators): Working with 1 authenticator(s)
debug(pam_u2f): util.c:426 (get_authenticators): Checking whether key exists in authenticator 0
debug(pam_u2f): util.c:436 (get_authenticators): Authenticator path: IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/GP17@7,1/IOPP/XHC0@0,3/XHC0@60000000/PRT3@60300000/USB2.0 Hub@60300000/AppleUSB20Hub@60300000/AppleUSB20HubPort@60310000/ePass FIDO@60310000/IOUSBHostInterface@0/IOUSBHostHIDDevice@60310000,0
debug(pam_u2f): util.c:463 (get_authenticators): Found key in authenticator 0
debug(pam_u2f): pam-u2f.c:391 (pam_sm_authenticate): do_authentication returned -2
debug(pam_u2f): pam-u2f.c:430 (pam_sm_authenticate): done. [authentication error]
~

**Comment by daysonhw** [comment]
`debug(pam_u2f): pam-u2f.c:105 (parse_cfg): called.
debug(pam_u2f): pam-u2f.c:106 (parse_cfg): flags 0 argc 3
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): argv[0]=debug
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): argv[1]=debug_file=/Users/dayson/.config/u2f/debug.log
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): argv[2]=authfile=/Users/dayson/.config/u2f/keys
debug(pam_u2f): pam-u2f.c:110 (parse_cfg): max_devices=0
debug(pam_u2f): pam-u2f.c:111 (parse_cfg): debug=1
debug(pam_u2f): pam-u2f.c:112 (parse_cfg): interactive=0
debug(pam_u2f): pam-u2f.c:113 (parse_cfg): cue=0
debug(pam_u2f): pam-u2f.c:114 (parse_cfg): nodetect=0
debug(pam_u2f): pam-u2f.c:115 (parse_cfg): userpresence=0
debug(pam_u2f): pam-u2f.c:116 (parse_cfg): userverification=0
debug(pam_u2f): pam-u2f.c:117 (parse_cfg): pinverification=0
debug(pam_u2f): pam-u2f.c:118 (parse_cfg): manual=0
debug(pam_u2f): pam-u2f.c:119 (parse_cfg): nouserok=0
debug(pam_u2f): pam-u2f.c:120 (parse_cfg): openasuser=0
debug(pam_u2f): pam-u2f.c:121 (parse_cfg): alwaysok=0
debug(pam_u2f): pam-u2f.c:123 (parse_cfg): authfile=/Users/dayson/.config/u2f/keys
debug(pam_u2f): pam-u2f.c:125 (parse_cfg): authpending_file=(null)
debug(pam_u2f): pam-u2f.c:126 (parse_cfg): origin=(null)
debug(pam_u2f): pam-u2f.c:127 (parse_cfg): appid=(null)
100 fewer lines
debug(pam_u2f): pam-u2f.c:366 (pam_sm_authenticate): Unable to emit 'authentication started' notification by opening the file '/var/run/user/0/pam-u2f-authpending', (No such file or directory)
debug(pam_u2f): util.c:527 (do_authentication): Device max index is 1
debug(pam_u2f): util.c:560 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:581 (do_authentication): Key handle: syF2PvdaZEupAook8N7X2lkEpxtPKDcceHxpNqvmKYZj2jajmHqKg9CcqCK1h4XcnUkJIQO2MugSj82E231CmouQPJzKRsvsNt/td1rnLgY==
debug(pam_u2f): util.c:673 (do_authentication): Challenge: ZdrTc7DYB8EhTV/ZnYMoXI3n6nIhxjoESnH9IY/oCOo=
debug(pam_u2f): util.c:422 (get_authenticators): Working with 1 authenticator(s)
debug(pam_u2f): util.c:426 (get_authenticators): Checking whether key exists in authenticator 0
debug(pam_u2f): util.c:436 (get_authenticators): Authenticator path: IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/GP17@7,1/IOPP/XHC0@0,3/XHC0@60000000/PRT3@60300000/USB2.0 Hub@60300000/AppleUSB20Hub@60300000/AppleUSB20HubPort@60310000/ePass FIDO@60310000/IOUSBHostInterface@0/IOUSBHostHIDDevice@60310000,0
debug(pam_u2f): util.c:463 (get_authenticators): Found key in authenticator 0
debug(pam_u2f): pam-u2f.c:391 (pam_sm_authenticate): do_authentication returned -2
debug(pam_u2f): pam-u2f.c:430 (pam_sm_authenticate): done. [authentication error]
debug(pam_u2f): pam-u2f.c:105 (parse_cfg): called.


debug(pam_u2f): pam-u2f.c:366 (pam_sm_authenticate): Unable to emit 'authentication started' notification by opening the file '/var/run/user/0/pam-u2f-authpending', (No such file or directory)
debug(pam_u2f): util.c:527 (do_authentication): Device max index is 1
debug(pam_u2f): util.c:560 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:581 (do_authentication): Key handle: syF2PvdaZEupAook8N7X2lkEpxtPKDcceHxpNqvmKYZj2jajmHqKg9CcqCK1h4XcnUkJIQO2MugSj82E231CmouQPJzKRsvsNt/td1rnLgY==
debug(pam_u2f): util.c:673 (do_authentication): Challenge: IR3YPMYjtlTDAM2QGHe2caq/zIsJCenAsZOJKz1Zi2s=
debug(pam_u2f): util.c:422 (get_authenticators): Working with 1 authenticator(s)
debug(pam_u2f): util.c:426 (get_authenticators): Checking whether key exists in authenticator 0
debug(pam_u2f): util.c:436 (get_authenticators): Authenticator path: IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/GP17@7,1/IOPP/XHC0@0,3/XHC0@60000000/PRT3@60300000/USB2.0 Hub@60300000/AppleUSB20Hub@60300000/AppleUSB20HubPort@60310000/ePass FIDO@60310000/IOUSBHostInterface@0/IOUSBHostHIDDevice@60310000,0
debug(pam_u2f): util.c:463 (get_authenticators): Found key in authenticator 0
debug(pam_u2f): pam-u2f.c:391 (pam_sm_authenticate): do_authentication returned -2
debug(pam_u2f): pam-u2f.c:430 (pam_sm_authenticate): done. [authentication error]
~`

**Comment by a-dma** [comment]
right, now you have to add one more flag to test for user presence. Try adding `,es256,+presence` at the end of the configuration line in your authfile (the one at `/Users/dayson/.config/u2f/keys`), or regenerate it using the newly built `pamu2fcfg`.

**Comment by daysonhw** [comment]
...because make error I passed to build "man", and I didn't get the new pamu2fcfg.
So, I added the code to the end of the line. same problem happened.
And the debug says
```
dayson:5VQoiZDwLdTMh7taHQff-ags9aAN44XvkykQnFfAh3_gkATeDvFCZBnm7OUMn6xAkfGbv951_

debug(pam_u2f): pam-u2f.c:282 (pam_sm_authenticate): Using authentication file /Users/dayson/.config/u2f/keys
debug(pam_u2f): util.c:215 (get_devices_from_authfile): Authorization line: dayson:5VQoiZDwLdTMh7taHQff-ags9aAN44XvkykQnFfAh3_gkATeDvFCZBnm7OUMn6xAkfGbv951_QAHOp7xQj-2D2fZWA7iiHFMlR9OViiMEK4,04cb62bf2a278d67ddaa958953243f4eab46d6b2a5c637e4d4e0a62faece46ef925c4eb7b63fd466cb0e549bfefae899885b70cee1c3b9446fd7ef5b41f4da8805
debug(pam_u2f): util.c:220 (get_devices_from_authfile): Matched user: dayson
debug(pam_u2f): util.c:256 (get_devices_from_authfile): KeyHandle for device number 1: 5VQoiZDwLdTMh7taHQff-ags9aAN44XvkykQnFfAh3_gkATeDvFCZBnm7OUMn6xAkfGbv951_QAHOp7xQj-2D2fZWA7iiHFMlR9OViiMEK4
debug(pam_u2f): util.c:279 (get_devices_from_authfile): publicKey for device number 1: 04cb62bf2a278d67ddaa958953243f4eab46d6b2a5c637e4d4e0a62faece46ef925c4eb7b63fd466cb0e549bfefae899885b70cee1c3b9446fd7ef5b41f4da8805
debug(pam_u2f): util.c:296 (get_devices_from_authfile): Unable to retrieve COSE type 1
debug(pam_u2f): util.c:297 (get_devices_from_authfile): Assuming ES256 (backwards compatibility)
debug(pam_u2f): util.c:318 (get_devices_from_authfile): Unable to retrieve attributes 1
debug(pam_u2f): util.c:319 (get_devices_from_authfile): Assuming 'p' (backwards compatibility)
debug(pam_u2f): util.c:215 (get_devices_from_authfile): Authorization line: ,es256,+presence
debug(pam_u2f): pam-u2f.c:105 (parse_cfg): called.
debug(pam_u2f): pam-u2f.c:106 (parse_cfg): flags -2147483648 argc 4
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): argv[0]=cue
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): argv[1]=debug
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): argv[2]=debug_file=/Users/dayson/.config/u2f/debug.log
debug(pam_u2f): pam-u2f.c:108 (parse_cfg): argv[3]=authfile=/Users/dayson/.config/u2f/keys
debug(pam_u2f): pam-u2f.c:110 (parse_cfg): max_devices=0
debug(pam_u2f): pam-u2f.c:111 (parse_cfg): debug=1
debug(pam_u2f): pam-u2f.c:112 (parse_cfg): interactive=0
debug(pam_u2f): pam-u2f.c:113 (parse_cfg): cue=1
debug(pam_u2f): pam-u2f.c:114 (parse_cfg): nodetect=0
debug(pam_u2f): pam-u2f.c:115 (parse_cfg): userpresence=0
debug(pam_u2f): pam-u2f.c:116 (parse_cfg): userverification=0
debug(pam_u2f): pam-u2f.c:117 (parse_cfg): pinverification=0
debug(pam_u2f): pam-u2f.c:118 (parse_cfg): manual=0
debug(pam_u2f): pam-u2f.c:119 (parse_cfg): nouserok=0
debug(pam_u2f): pam-u2f.c:120 (parse_cfg): openasuser=0
debug(pam_u2f): pam-u2f.c:121 (parse_cfg): alwaysok=0
debug(pam_u2f): pam-u2f.c:123 (parse_cfg): authfile=/Users/dayson/.config/u2f/keys
debug(pam_u2f): pam-u2f.c:125 (parse_cfg): authpending_file=(null)
debug(pam_u2f): pam-u2f.c:126 (parse_cfg): origin=(null)
debug(pam_u2f): pam-u2f.c:366 (pam_sm_authenticate): Unable to emit 'authentication started' notification by opening the file '/var/run/user/0/pam-u2f-authpending', (No such file or directory)
debug(pam_u2f): util.c:527 (do_authentication): Device max index is 1
debug(pam_u2f): util.c:560 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:581 (do_authentication): Key handle: 5VQoiZDwLdTMh7taHQff+ags9aAN44XvkykQnFfAh3/gkATeDvFCZBnm7OUMn6xAkfGbv951/QAHOp7xQj+2D2fZWA7iiHFMlR9OViiMEK4==
debug(pam_u2f): util.c:673 (do_authentication): Challenge: vOV1urezhsPbAJu2KfvmF+awq12wdxmqgraHD+QRns4=
debug(pam_u2f): util.c:422 (get_authenticators): Working with 1 authenticator(s)
debug(pam_u2f): util.c:426 (get_authenticators): Checking whether key exists in authenticator 0
debug(pam_u2f): util.c:436 (get_authenticators): Authenticator path: IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/GP17@7,1/IOPP/XHC0@0,3/XHC0@60000000/PRT3@60300000/USB2.0 Hub@60300000/AppleUSB20Hub@60300000/AppleUSB20HubPort@60310000/ePass FIDO@60310000/IOUSBHostInterface@0/IOUSBHostHIDDevice@60310000,0
debug(pam_u2f): util.c:463 (get_authenticators): Found key in authenticator 0
debug(pam_u2f): pam-u2f.c:391 (pam_sm_authenticate): do_authentication returned -2
debug(pam_u2f): pam-u2f.c:430 (pam_sm_authenticate): done. [authentication error]
~

debug(pam_u2f): pam-u2f.c:366 (pam_sm_authenticate): Unable to emit 'authentication started' notification by opening the file '/var/run/user/0/pam-u2f-authpending', (No such file or directory)
debug(pam_u2f): util.c:527 (do_authentication): Device max index is 1
debug(pam_u2f): util.c:560 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:581 (do_authentication): Key handle: 5VQoiZDwLdTMh7taHQff+ags9aAN44XvkykQnFfAh3/gkATeDvFCZBnm7OUMn6xAkfGbv951/QAHOp7xQj+2D2fZWA7iiHFMlR9OViiMEK4==
debug(pam_u2f): util.c:673 (do_authentication): Challenge: aiDxfDjSSXUIaymWYR+WXCC0nZRt/iFcjHyjjcrFpVE=
debug(pam_u2f): util.c:422 (get_authenticators): Working with 1 authenticator(s)
debug(pam_u2f): util.c:426 (get_authenticators): Checking whether key exists in authenticator 0
debug(pam_u2f): util.c:436 (get_authenticators): Authenticator path: IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/GP17@7,1/IOPP/XHC0@0,3/XHC0@60000000/PRT3@60300000/USB2.0 Hub@60300000/AppleUSB20Hub@60300000/AppleUSB20HubPort@60310000/ePass FIDO@60310000/IOUSBHostInterface@0/IOUSBHostHIDDevice@60310000,0
debug(pam_u2f): util.c:463 (get_authenticators): Found key in authenticator 0
debug(pam_u2f): pam-u2f.c:391 (pam_sm_authenticate): do_authentication returned -2
debug(pam_u2f): pam-u2f.c:430 (pam_sm_authenticate): done. [authentication error]
~
```

**Comment by a-dma** [comment]
You may have to regenerate the credential with `pamu2fcfg`, it looks like the library is not happy with it.

You can build without man pages (and all the dependencies that they requires) with `./configure --disable-man`.

**Comment by Metang386** [comment]
Same issue, tried all of the above and none of these suggestions are helping.  Where can I download Pam 1.0.4?

**Comment by wclarie** [comment]
TLDR; HEAD works on Mojave!

OK I ran into the same issue on Mojave (See also #128) with the latest release of pam-u2f: my keys (Yubikey U2F) working fine with pamu2fcfg, but then failing to work with the actual pam module (I'm testing with sudo).

I just built pam-u2f HEAD the following way:

Make sure the old version is no longer lying around:

> brew uninstall pam-u2f

Install libcbor (dependency) from a submitted Homebrew formula which got rejected (?):

> brew install https://raw.githubusercontent.com/Homebrew/homebrew-core/02ca54688cf404d5e268370629ab183f6eb8fb47/Formula/libcbor.rb

Install libfido2 (dependency) from a submitted Homebrew formula which got rejected as well (?):

> brew install https://raw.githubusercontent.com/gvarisco/homebrew-core/patch-2/Formula/libfido2.rb

Build the pam module (you may need some build dependencies):

> git clone https://github.com/Yubico/pam-u2f.git
> cd pam-u2f
> autoreconf --install
> export PKG_CONFIG_PATH=/usr/local/opt/openssl@1.1/lib/pkgconfig
> ./configure --with-pam-dir=/usr/local/lib/pam/
> make
> make install

Then rebuild the keys file using the new version of `pamu2fcfg` as usual, and set up the pam config to point to the new module in `/usr/local/lib/pam/` and that should just work.

**Comment by a-dma** [comment]
a new version is out with updated dependencies. Closing.
